src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
         return BlockFreeSet{
-            .free = try std.bit_set.DynamicBitSetUnmanaged.initFull(count, allocator),
+            .index = index,
+            .blocks = blocks,
+            .recent = .{},
         };
     }
 
     pub fn deinit(set: *BlockFreeSet, allocator: *mem.Allocator) void {
-        set.free.deinit(allocator);
+        set.index.deinit(allocator);
+        set.blocks.deinit(allocator);
     }
 
-    // TODO consider "caching" the first set bit to speed up subsequent acquire() calls
-    pub fn acquire(set: *BlockFreeSet) u64 {
-        // TODO: To ensure this "unreachable" is never reached, the leader must reject
-        // new requests when storage space is too low to fulfill them.
-        const bit = set.free.findFirstSet() orelse unreachable;
-        set.free.unset(bit);
-        const address = bit + 1;
+    /// Marks a free block as allocated, and returns the address. Panics if no blocks are available.
+    pub fn acquire(set: *BlockFreeSet) ?u64 {
+        const block = blk: {
+            if (set.recent.pop()) |block| {
+                break :blk block;
+            } else if (set.index.findFirstSet()) |shard| {
+                break :blk set.find_free_block_in_shard(shard) orelse unreachable;
+            } else return null;
+        };
+        const shard = block / shard_size;
+        assert(set.blocks.isSet(block));
+        assert(set.index.isSet(shard));
+
+        set.blocks.unset(block);
+        // Update the index when every block in the shard is allocated.
+        if (set.find_free_block_in_shard(shard) == null) set.index.unset(shard);
+
+        const address = block + 1;
         return @intCast(u64, address);
     }
 
+    fn find_free_block_in_shard(set: *BlockFreeSet, shard: usize) ?usize {
+        const shard_start = shard * shard_size;
+        const shard_end = shard_start + shard_size;
+        assert(shard_start < set.blocks.bit_length);
+
+        return find_first_set_bit(set.blocks, shard_start, shard_end);
+    }
+
+    fn is_free(set: *BlockFreeSet, address: u64) bool {
+        const block = address - 1;
+        return set.blocks.isSet(block);
+    }
+
+    /// Marks the specified block as free.
     pub fn release(set: *BlockFreeSet, address: u64) void {
-        const bit = address - 1;
-        assert(!set.free.isSet(bit));
-        set.free.set(bit);
+        const block = address - 1;
+        assert(!set.blocks.isSet(block));
+
+        set.index.set(block / shard_size);
+        set.blocks.set(block);
+        set.recent.push(block) catch {}; // ignore error.NoSpaceLeft
+    }
+
+    /// Decodes the compressed bitset in `source` into `set`. Panics if `source`'s encoding is invalid.
+    pub fn decode(set: *BlockFreeSet, source: []align(@alignOf(usize)) const u8) void {
+        // Verify that this BlockFreeSet is entirely unallocated.
+        assert(set.index.count() == set.index.bit_length);
+
+        _ = EWAH.decode(source, bitset_masks(set.blocks));
```

Comment:
What do we expect of the return value? For example, if we know that `source.len > @sizeOf(Word)` then we should expect some data to have been decoded, and should be able to cross-check this again against `set.index.bit_length` in some way (e.g. that it's not completely full).

Reply:
I can add:

```
assert(words_decoded * @bitSizeOf(MaskInt) <= set.blocks.bit_length);
```

The reason it's `<=` instead of `==` is that the API deliberately allows allocating a larger bitset than the one being decoded, to allow the `BlockFreeSet` to grow (e.g. if the database is reconfigured).

This condition is implicit, since `EWAH.decode` checks it and also if it were false then it would have already failed anyway (trying to write to the memory out of bounds). I can add the assertion anyway for clarity though.